engine/src/main/java/org/hibernate/validator/internal/properties/javabean/JavaBeanHelper.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.properties.javabean;
+
+import static org.hibernate.validator.internal.util.logging.Messages.MESSAGES;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Optional;
+
+import org.hibernate.validator.internal.properties.ConstrainableType;
+import org.hibernate.validator.internal.util.Contracts;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredConstructor;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredField;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredMethod;
+import org.hibernate.validator.internal.util.privilegedactions.GetMethod;
+import org.hibernate.validator.spi.properties.ConstrainableExecutable;
+import org.hibernate.validator.spi.properties.GetterPropertySelectionStrategy;
+
+/**
+ * {@link ConstrainableType} that describes a JavaBeanHelper.
+ *
+ * @author Marko Bekhta
+ */
+public class JavaBeanHelper {
+
+	private final GetterPropertySelectionStrategy getterPropertySelectionStrategy;
+
+	public JavaBeanHelper(GetterPropertySelectionStrategy getterPropertySelectionStrategy) {
+		this.getterPropertySelectionStrategy = getterPropertySelectionStrategy;
+	}
+
+	public Optional<JavaBeanField> findDeclaredFieldProperty(Class<?> declaringClass, String property) {
+		Contracts.assertNotNull( declaringClass, MESSAGES.classCannotBeNull() );
+
+		Field field = run( GetDeclaredField.action( declaringClass, property ) );
+		return Optional.ofNullable( field ).map( JavaBeanField::new );
+	}
+
+	public Optional<JavaBeanGetter> findDeclaredGetterProperty(Class<?> declaringClass, String property) {
+		Contracts.assertNotNull( declaringClass, MESSAGES.classCannotBeNull() );
+		for ( String possibleName : getterPropertySelectionStrategy.getPossibleMethodNamesForGetterProperty( property ) ) {
+			Method getter = run( GetDeclaredMethod.action( declaringClass, possibleName ) );
```

Comment:
I think it would be better to only call one secured method, similar to what we had before. The big difference would be you would pass the list of candidates to the action.